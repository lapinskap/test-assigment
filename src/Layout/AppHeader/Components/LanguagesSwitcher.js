import React, { useState } from 'react';
import {
  DropdownItem, DropdownMenu, DropdownToggle, UncontrolledDropdown,
} from 'reactstrap';
import Flag from 'react-flagkit';

const codeFlagMap = {
  pl: 'PL',
  en: 'GB',
  uk: 'UA',
};

export default function LanguagesSwitcher() {
  const [lang, setLang] = useState('en');
  const languages = ['pl', 'en'];
  if (languages.length <= 1) {
    return null;
  }

  const changeLanguage = (key) => {
    setLang(key);
    window.location.reload();
  };

  // const currentLanguage = languages.find((el) => el.code === lang);
  return (
    <UncontrolledDropdown data-t1="languageSwitcher">
      <DropdownToggle className="p-0 mr-2" color="link">
        <div className="icon-wrapper icon-wrapper-alt rounded-circle">
          <div className="icon-wrapper-bg bg-focus" />
          <div className="language-icon">
            <Flag className="mr-3 opacity-8" country="JE" size="40" />
          </div>
        </div>
      </DropdownToggle>
      <DropdownMenu right className="rm-pointers">
        <div className="dropdown-menu-header">
          <div className="dropdown-menu-header-inner pt-4 pb-4 bg-focus">
            <div className="menu-header-image opacity-05" />
            <div className="menu-header-content text-center text-white">
              <h6 className="menu-header-subtitle mt-0">
                Select language
              </h6>
            </div>
          </div>
        </div>
        {languages.map(({ code, label }) => (
          <DropdownItem key={code} active={code === lang} onClick={() => changeLanguage(code)}>
            <Flag className="mr-3 opacity-8" country={codeFlagMap[code] || 'JE'} />
            {label}
          </DropdownItem>
        ))}
      </DropdownMenu>
    </UncontrolledDropdown>
  );
}
