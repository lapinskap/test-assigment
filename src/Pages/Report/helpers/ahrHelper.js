export const getAhrUrl = (url, isAhr) => {
  let result = isAhr ? '/ahr' : '';
  result += url;

  return result;
};

export const asd = () => {};
