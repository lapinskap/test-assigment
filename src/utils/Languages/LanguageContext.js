import React, {
  useContext, useEffect, useState,
} from 'react';
import { defaultLanguage } from '../Translations/translationUtils';

const LanguageContext = React.createContext({
  languagesConfig: [],
  defaultLanguage: '',
});

export const LANGUAGE_STORAGE_KEY = 'user_language';

export const getLanguage = () => {
  const value = window.localStorage.getItem(LANGUAGE_STORAGE_KEY);
  if (value) {
    return value;
  }
  return defaultLanguage;
};
export const setLanguage = (lang) => {
  window.localStorage.setItem(LANGUAGE_STORAGE_KEY, lang);
};
export const removeLanguage = () => {
  window.localStorage.removeItem(LANGUAGE_STORAGE_KEY);
};

export const useLanguages = (withDefault = false, asOptions = false) => {
  const [result, setResult] = useState([]);
  const { languagesConfig } = useContext(LanguageContext);
  useEffect(() => {
    if (withDefault) {
      setResult(languagesConfig);
    } else {
      setResult(languagesConfig.filter(({ isDefault }) => !isDefault));
    }
  }, [withDefault, languagesConfig]);
  if (!asOptions) {
    return result;
  }
  return result.map(({ code, label }) => ({ value: code, label }));
};

export const useDefaultLanguage = () => {
  const { defaultLanguage: result } = useContext(LanguageContext);

  return result;
};

export default LanguageContext;
