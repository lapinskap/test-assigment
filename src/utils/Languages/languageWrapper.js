import React, { useContext, useEffect, useState } from 'react';
import PropTypes from 'prop-types';
import pl from 'date-fns/locale/pl';
import en from 'date-fns/locale/en-US';
import uk from 'date-fns/locale/uk';
import { registerLocale, setDefaultLocale } from 'react-datepicker';
import { COMPANY_MANAGEMENT_SERVICE, restApiRequest, TRANSLATOR_SERVICE } from '../Api';
import { INTERFACE_CODE } from '../Translations/translationUtils';
import LanguageContext, { setLanguage, getLanguage } from './LanguageContext';
import DefaultFallback from '../../Layout/AppMain/DefaultFallback';
import RbsContext from '../RoleBasedSecurity/RbsContext';

export const mockLanguages = [
  {
    code: 'pl',
    label: 'polski',
    isDefault: true,
    shortLabel: 'pl',
  },
  {
    code: 'en',
    label: 'angielski',
    isDefault: false,
    shortLabel: 'en',
  },
  // {
  //   code: 'uk',
  //   label: 'ukraiÅ„ski',
  //   isDefault: false,
  //   shortLabel: 'UK',
  // },
];

export const getLanguagesConfig = async () => {
  try {
    return await restApiRequest(
      TRANSLATOR_SERVICE,
      '/languages',
      'GET',
      {},
      mockLanguages,
    );
  } catch (e) {
    return [...mockLanguages];
  }
};
export const getCompanyLanguages = async (companyId) => {
  try {
    const response = await restApiRequest(
      COMPANY_MANAGEMENT_SERVICE,
      `/companies/${companyId}`,
      'GET',
      {},
      [],
    );
    return response?.availableLanguages || [];
  } catch (e) {
    return [];
  }
};

export const loadInterfaceTranslation = async (language) => {
  try {
    const response = await restApiRequest(
      TRANSLATOR_SERVICE,
      '/get-by-scope',
      'GET',
      {
        params: {
          language,
          value: [INTERFACE_CODE],
        },
      },
      '{}',
    );
    window.translations = JSON.parse(response);
  } catch (e) {
    console.error(e);
  }
};

const init = async (setTranslationsLoaded, setLanguagesConfig, setDefaultLanguage, userInfo) => {
  const promises = [getLanguagesConfig()];

  const [allSystemLanguages, companyLanguages = null] = await Promise.all(promises);
  const languageConfig = allSystemLanguages;

  const defaultLanguage = languageConfig.find(({ isDefault }) => isDefault);
  let language = getLanguage();

  if (!languageConfig.find(({ code }) => code === language)) {
    language = defaultLanguage.code;
    setLanguage(language);
  }
  registerGlobalLocale(language);
  if (defaultLanguage.code !== language) {
    await loadInterfaceTranslation(language);
  }
  setDefaultLanguage(defaultLanguage);

  setLanguagesConfig(languageConfig);
  setTranslationsLoaded(true);
};

export default function LanguageWrapper({ children }) {
  const [translationsLoaded, setTranslationsLoaded] = useState(false);
  const [languagesConfig, setLanguagesConfig] = useState({});
  const [defaultLanguage, setDefaultLanguage] = useState(null);
  const { userInfo } = useContext(RbsContext);

  useEffect(() => {
    init(setTranslationsLoaded, setLanguagesConfig, setDefaultLanguage, userInfo);
  }, [setTranslationsLoaded, setLanguagesConfig, setDefaultLanguage, userInfo]);

  if (!translationsLoaded) {
    return DefaultFallback;
  }
  return (
    <LanguageContext.Provider value={{
      languagesConfig,
      defaultLanguage,
    }}
    >
      {children}
    </LanguageContext.Provider>
  );
}

const registerGlobalLocale = (languageCode) => {
  const locale = languageToLocaleMap[languageCode] || pl;
  registerLocale(languageCode, locale); // register it with the name you want
  setDefaultLocale(languageCode);
};

const languageToLocaleMap = {
  pl,
  en,
  uk,
};

LanguageWrapper.propTypes = {
  children: PropTypes.node.isRequired,
};
