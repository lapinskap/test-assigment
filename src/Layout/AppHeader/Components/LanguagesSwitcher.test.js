import React from 'react';
import { mount } from 'enzyme';
import renderer, { act } from 'react-test-renderer';
import LanguagesSwitcher from './LanguagesSwitcher';
import '../../../tests/setupTests';
import TestAppContext from '../../../tests/TestAppContext';
import { wait } from '../../../tests/helpers';
import { removeLanguage } from '../../../utils/Languages/LanguageContext';

global.document.createRange = () => ({
  setStart: () => {},
  setEnd: () => {},
  commonAncestorContainer: {
    nodeName: 'BODY',
    ownerDocument: document,
  },
});

describe('<Languages />', () => {
  const setState = jest.fn();
  const useStateSpy = jest.spyOn(React, 'useState');
  useStateSpy.mockImplementation((init) => [init, setState]);

  afterEach(() => {
    jest.clearAllMocks();
    removeLanguage();
  });

  it('renders correctly', () => {
    const tree = renderer
      .create(
        <TestAppContext rbs language>
          <LanguagesSwitcher />
        </TestAppContext>,
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('should have correct behavior on all buttons', async () => {
    const wrapper = mount(
      <TestAppContext rbs language>
        <LanguagesSwitcher />
      </TestAppContext>,
    );
    await wait(200);
    wrapper.update();
    const langFlag = wrapper.find('Flag').at(1);
    expect(langFlag.props().value).toBe(undefined);
    await act(async () => {
      await langFlag.simulate('click');
    });
    wrapper.update();
    const DropdownItem = wrapper.find('DropdownItem').at(0);
    act(() => {
      DropdownItem.simulate('click');
    });
  });

  it('should changing language method work', async () => {
    const wrapper = mount(
      <TestAppContext rbs language>
        <LanguagesSwitcher />
      </TestAppContext>,
    );
    await wait(200);
    wrapper.update();
    const DropdownItem = wrapper.find('DropdownItem').at(0);
    act(() => {
      DropdownItem.simulate('click');
    });
    expect(DropdownItem.props().value).toBe(undefined);
  });
});
