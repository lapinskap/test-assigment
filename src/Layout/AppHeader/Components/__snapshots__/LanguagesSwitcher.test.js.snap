// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Languages /> renders correctly 1`] = `
<div
  className="loader-container"
>
  <div
    className="loader-container-inner"
  >
    <div
      className="text-center"
    >
      <div
        className="loader loader-undefined loader-active"
      >
        <div
          className="loader-inner line-scale-party"
        >
          <div
            style={Object {}}
          />
          <div
            style={Object {}}
          />
          <div
            style={Object {}}
          />
          <div
            style={Object {}}
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
